### PR TITLE
gpcheckperf: Fix memory calculation function

### PR DESCRIPTION
**Pull request by jnihal**
Currently, there is a bug in the `getMemory()` function in `gpcheckperf` because of the way we check the return code of the `run()` method which is called inside `getMemory()`. The `run()` method returns an integer value of `zero` in case of success and a `non-zero` value if it fails.

We are checking this value using the condition `if not ok` which is incorrect because when the `run()` method succeeds (`ok = 0`), the condition would result as `False` causing the `getMemory()` function to assume that the `run()` method failed but in reality, it did not.

A simple fix would be to change the condition from `if not ok` to `if ok != 0` to check for any possible failure from the `run()` method.

Further, the way `getMemory()` handles errors is also incorrect. It just returns `0` whenever there is an error which can lead to the incorrect file size being used to perform disk performance tests. This is because the gpcheckperf utility internally calls `multidd` command to perform disk performance tests which also accepts a file size parameter with the -S option which when not set (or is equal to 0), uses its own default value of `2 * memory_size`, but instead of dividing it from the number of input directories, it uses this value for each directory i.e. the total file size value would be `2 * memory_size * no_of_input_directories`. Due to this, sometimes the user can meet File System Full error when the number of input directories is big.

Hence we need to properly handle the error in the `getMemory()` function and exit from the code instead of just returning `0`

Before:
```
$ gpcheckperf -d /tmp/test1 -d /tmp/test2 -d /tmp/test3 -h localhost -rd
 disk write avg time (sec): 5.46
 disk write tot bytes: 12884901888     --> is equal to 2 * memory_size * no_of_input_directories
 disk write tot bandwidth (MB/s): 2250.55
 disk write min bandwidth (MB/s): 2250.55 [localhost]
 disk write max bandwidth (MB/s): 2250.55 [localhost]
 ```
After:
```
$ gpcheckperf -d /tmp/test1 -d /tmp/test2 -d /tmp/test3 -h localhost -rd
 disk write avg time (sec): 1.87
 disk write tot bytes: 4295000064     --> is equal to 2 * memory_size
 disk write tot bandwidth (MB/s): 2190.39
 disk write min bandwidth (MB/s): 2190.39 [localhost]
 disk write max bandwidth (MB/s): 2190.39 [localhost]
```

Also added unit test cases to test the getMemory() function outputs and added a main section to the gpcheckperf code.

[Pipeline](https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev-7x_gpcheckperf_mem_fix)

### CODE REVIEW

**Comment by shrakesh** [comment]
Please elaborate more on the PR description and provide an example to better understand the function output(type) before and after the changes made in the code.

**Comment by ravoorsh** [comment]
> Please elaborate more on the PR description and provide an example to better understand the function output(type) before and after the changes made in the code.

+1

File: gpMgmt/bin/gpcheckperf [hunk 1]
```diff
@@ -135,24 +135,26 @@ def getPlatform():
...
-            return 0
+        if ok != 0:
+            return -1
         word_list = out.strip().split(' ')
         val = int(word_list[len(word_list) - 2])
         factor = word_list[len(word_list) - 1]
         if factor == 'kB':
             return val * 1024
-        return 0
+        else:
```
Comment (shrakesh) [inline gpMgmt/bin/gpcheckperf:145]: duplicate code(line 145-146 and line 156-157) can be avoided and we can have a default return at the end of the function.

File: gpMgmt/bin/gpcheckperf [hunk 2]
```diff
@@ -135,24 +135,26 @@ def getPlatform():
 def getMemory():
     if getPlatform() == 'linux':
```
Comment (shrakesh) [inline gpMgmt/bin/gpcheckperf:136]: can we store output from getPlatform() and use that it will avoid calling twice the same function?

File: gpMgmt/bin/gpcheckperf [hunk 3]
```diff
@@ -135,24 +135,26 @@ def getPlatform():
...
         val = int(word_list[len(word_list) - 2])
         factor = word_list[len(word_list) - 1]
         if factor == 'kB':
             return val * 1024
-        return 0
+        else:
+            return -1
 
-    if getPlatform() == 'darwin':
+    elif getPlatform() == 'darwin':
```
Comment (shrakesh) [inline gpMgmt/bin/gpcheckperf:148]: any specific reason to replace if with elif ?

File: gpMgmt/bin/gpcheckperf [hunk 4]
```diff
@@ -135,24 +135,26 @@ def getPlatform():
...
         ok, out = run("/usr/sbin/sysctl hw.physmem")
-        if not ok:
-            return 0
+        if ok != 0:
+            return -1
```
Comment (shrakesh) [inline gpMgmt/bin/gpcheckperf:151]: IMO it would be good to return None instead of -1.

File: gpMgmt/bin/gpcheckperf [hunk 5]
```diff
@@ -261,7 +263,12 @@ def parseCommandLine():
         usage('Error: maximum size for -B parameter is 1MB')
 
     if GV.opt['-S'] == 0:
-        GV.opt['-S'] = 2 * getMemory() / len(GV.opt['-d'])
+        system_mem_size = getMemory()
+        if system_mem_size == -1:
```
Comment (shrakesh) [inline gpMgmt/bin/gpcheckperf:267]: In some hypothetical corner case, lets say getMemory() returns 0.  as we are checking for -1, and in the else part, it will become the scene of the original issue. I think checking Not None would benefit here.

File: gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckperf.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+        mock_run.return_value = [0, 'hw.physmem: 1234']
+        actual_result = self.subject.getMemory()
+        self.assertEquals(actual_result, 1234)
+
+    @patch('gpcheckperf.getPlatform', return_value='linux')
+    @patch('gpcheckperf.run')
+    def test_get_memory_on_linux(self, mock_run, mock_get_platform):
+        mock_run.return_value = [1, 'MemTotal:        10 kB']
+        actual_result = self.subject.getMemory()
```
Comment (shrakesh) [inline gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckperf.py:29]: we can also add a corner case of MemTotal returning 0kB on successful command execution.

File: gpMgmt/bin/gpcheckperf [hunk 7]
```diff
@@ -261,7 +263,12 @@ def parseCommandLine():
...
+            sys.exit('[Error] Unable to get system memory size. Instead, you can use the -S option to provide the file size value.')
```
Comment (shrakesh) [inline gpMgmt/bin/gpcheckperf:268]: can we cover these newly added workflows to behave test case? we should add some for gpcheckperf -S option as well.

**Review by shrakesh — CHANGES_REQUESTED** [review]

File: gpMgmt/bin/gpcheckperf (hunk 1 above)
Comment (jnihal) [reply gpMgmt/bin/gpcheckperf]: Added a default return value at the end.

File: gpMgmt/bin/gpcheckperf (hunk 3 above)
Comment (jnihal) [reply gpMgmt/bin/gpcheckperf]: I added an else part at the end, so made the change. But now since we are returning a default value there is no need for it. Will revert it back.

File: gpMgmt/bin/gpcheckperf (hunk 4 above)
Comment (jnihal) [reply gpMgmt/bin/gpcheckperf]: Agreed, returning a `None` would be better.

File: gpMgmt/bin/gpcheckperf (hunk 5 above)
Comment (jnihal) [reply gpMgmt/bin/gpcheckperf]: Made a change in the getMemory() function so that it never returns a 0 value which will make sure the issue is avoided

File: gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckperf.py (hunk 6 above)
Comment (jnihal) [reply gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckperf.py:33]: Added the test

File: gpMgmt/bin/gpcheckperf [hunk 8]
```diff
@@ -133,26 +133,27 @@ def getPlatform():
...
 
 def getMemory():
-    if getPlatform() == 'linux':
+    platform_type = getPlatform()
+
+    if platform_type == 'linux':
         ok, out = run("sh -c 'cat /proc/meminfo | grep MemTotal'")
-        if not ok:
-            return 0
+        if ok != 0:
```
Comment (shrakesh) [inline gpMgmt/bin/gpcheckperf:140]: can we change ok to rc as this indicates the return code from the command?

File: gpMgmt/bin/gpcheckperf [hunk 9]
```diff
@@ -261,7 +262,12 @@ def parseCommandLine():
         usage('Error: maximum size for -B parameter is 1MB')
 
     if GV.opt['-S'] == 0:
-        GV.opt['-S'] = 2 * getMemory() / len(GV.opt['-d'])
+        system_mem_size = getMemory()
+        if system_mem_size != None:
```
Comment (shrakesh) [inline gpMgmt/bin/gpcheckperf:266]: more preferred way 
if system_mem_size is not None:

**Review by shrakesh — APPROVED** [review]
Approving with some minor suggestions.

File: gpMgmt/bin/gpcheckperf [hunk 10]
```diff
@@ -261,7 +262,12 @@ def parseCommandLine():
...
+            GV.opt['-S'] = 2 * system_mem_size / len(GV.opt['-d'])
+        else:
+            sys.exit('[Error] Unable to get system memory size. Instead, you can use the -S option to provide the file size value.')
```
Comment (Annu149) [inline gpMgmt/bin/gpcheckperf:269]: Shouldn't we assert this error message in behave testcase and unit testcase?

File: gpMgmt/test/behave/mgmt_utils/gpcheckperf.feature [hunk 11]
```diff
@@ -52,3 +52,8 @@ Feature: Tests for gpcheckperf
     And   gpcheckperf should print "single host only - abandon netperf test" to stdout
     And gpcheckperf should not print "TypeError:" to stdout
 
+  Scenario: gpcheckperf runs with -S option and prints a warning message
```
Comment (Annu149) [inline gpMgmt/test/behave/mgmt_utils/gpcheckperf.feature:55]: What if we provide an invalid value with -S? I guess parseMemorySize() function will throw an exception. We should add that scenario as well

File: gpMgmt/test/behave/mgmt_utils/gpcheckperf.feature [hunk 12]
```diff
@@ -52,3 +52,8 @@ Feature: Tests for gpcheckperf
...
+    Given the database is running
+    When  the user runs "gpcheckperf -h localhost -rd -d /tmp -S 1GB"
```
Comment (Annu149) [inline gpMgmt/test/behave/mgmt_utils/gpcheckperf.feature:57]: we should give space between -r and d for better understanding "gpcheckperf -h localhost -r d -d /tmp -S 1GB"

File: gpMgmt/bin/gpcheckperf (hunk 10 above)
Comment (ravoorsh) [reply gpMgmt/bin/gpcheckperf:269]: Please follow the Postgres guidelines for error messages 

Primary error messages: Do not capitalize the first letter. Do not end a message with a period. Do not even think about ending a message with an exclamation point.

Consider using "could not" instead of Unable

**Review by ravoorsh — CHANGES_REQUESTED** [review]

File: gpMgmt/test/behave/mgmt_utils/gpcheckperf.feature (hunk 11 above)
Comment (jnihal) [reply gpMgmt/test/behave/mgmt_utils/gpcheckperf.feature:55]: Added the scenario

File: gpMgmt/bin/gpcheckperf (hunk 10 above)
Comment (jnihal) [reply gpMgmt/bin/gpcheckperf]: > Shouldn't we assert this error message in behave testcase and unit testcase?

It is not possible to write a behave test case, because this error occurs only when the system memory size would be equal to 0.

**Review by Annu149 — APPROVED** [review]
Looks good to me

**Review by ravoorsh — APPROVED** [review]

**Review by jmcatamney — APPROVED** [review]

**Comment by shrakesh** [comment]
merged on behalf of nihal.
